sphinx/source/docs/user_guide/data.rst

```diff
@@ -25,69 +30,122 @@ to the ``circle`` plotting method (see :ref:`userguide_plotting` for more exampl
+the object's ``data`` parameter:
+
+* Bokeh uses the dictionary's keys as column names.
+* The dictionary's values are used as the data values for your ColumnDataSource.
+
+The data you pass as part of your dict can be any non-string ordered sequences
+of values, such as lists or arrays (including NumPy arrays).
+
+Once you have a ColumnDataSource set up, pass it to a plotting function using
+the ``source`` argument:
 
 .. code-block:: python
 
     from bokeh.plotting import figure
     from bokeh.models import ColumnDataSource
 
+    # create a Python dict as the basis of your ColumnDataSource
     data = {'x_values': [1, 2, 3, 4, 5],
             'y_values': [6, 7, 2, 3, 6]}
 
+    # create a ColumnDataSource by passing the dict
     source = ColumnDataSource(data=data)
 
+    # create a plot using the ColumnDataSource's two columns
     p = figure()
     p.circle(x='x_values', y='y_values', source=source)
 
+To use a ColumnDataSource with a plotting function, you need to pass at least
+these three arguments:
+
+* ``x``: the name of the ColumnDataSource's column that contains the data for
+  the x values of your plot
+* ``y``: the name of the ColumnDataSource's column that contains the data for
+  the y values of your plot
+* ``source``: the name of the ColumnDataSource that contains the columns you
+  just referenced for the ``x`` and ``y`` arguments.
+
 .. note::
-    There is an implicit assumption that all the columns in a given ``ColumnDataSource``
-    all have the same length at all times. For this reason, it is usually preferable to
-    update the ``.data`` property of a data source "all at once".
+    Bokeh assumes that all columns in a ColumnDataSource each have the same
+    length at all times. For this reason, make sure to always update all columns
+    of a ColumnDataSource at the same time.
 
+.. _userguide_data_cds_pandas_data_frame:
 
-Pandas
-~~~~~~
+Using a pandas DataFrame
+~~~~~~~~~~~~~~~~~~~~~~~~
 
-The ``data`` parameter can also be a Pandas ``DataFrame`` or ``GroupBy`` object.
+The ``data`` parameter can also be a pandas ``DataFrame`` or ``GroupBy`` object:
 
 .. code-block:: python
 
    source = ColumnDataSource(df)
 
-If a ``DataFrame`` is used, the CDS will have columns corresponding to the columns of
-the ``DataFrame``. The index of the ``DataFrame`` will be reset, so if the ``DataFrame``
-has a named index column, then CDS will also have a column with this name. However,
-if the index name is ``None``, then the CDS will be assigned a generic name.
-It will be ``index`` if it is available, and ``level_0`` otherwise.
+If you use a pandas ``DataFrame``, the resulting ColumnDataSource in Bokeh will
+have columns that correspond to the columns of the ``DataFrame``. The naming of
+the columns follows these rules:
+
+* The index of the ``DataFrame`` will be reset, so if the ``DataFrame`` has a
+  named index column, the ColumnDataSource will also have a column with this
+  name.
```

Comment:
This sounds like we are mutating the the input df by calling `reset_index` on it, but I don't think we do that?

Reply:
I was indeed unsure about what "The index of the ``DataFrame`` will be reset" in the original text means. Can you describe what is actually happening in this case? Would it be possible to just leave out this one sentence and go with:

> * If the ``DataFrame`` has a named index column, the ColumnDataSource will also have a column with this
  name.

Reply:
Yes I think that is perfect

Reply:
If you *really* want to dig in to the behaviour, the tests are here:

https://github.com/bokeh/bokeh/blob/branch-2.3/tests/unit/bokeh/models/test_sources.py#L73-L175

Reply:
And actually escpially relevant a couple more here:

https://github.com/bokeh/bokeh/blob/branch-2.3/tests/unit/bokeh/models/test_sources.py#L303-L329